src/Tasks/SystemState.cs

```diff
@@ -366,15 +366,6 @@ private FileState ComputeFileStateFromCachesAndDisk(string path)
             // If the process-wide cache contains an up-to-date FileState, always use it
             if (isProcessFileStateUpToDate)
             {
```

Comment:
Hmm, on the hand I think it is still slightly sub-optimal. Currently it is still indeterministic. For example imagine m=1 build with identical project A and B. If project A is build first cache for project B will be empty because process wide cache will be filled by project A.
I would recommend to have flow like:
``` 
IF in-process-cache
  IF not in-instance-cache
    update-instance-cache(cached-item)
  return cached-item
```

Reply:
Please consider to remeasure perf after above changes as it might make amount of serialized data considerably bigger. But even if it is slightly slower I would still prefer correctness and deterministic behavior.

Reply:
Even though our build graph is deterministic, the actual execution plan isn't and projects are build on more less random node. Additionally VS does not build solution but just projects, so if only inputs of projects B changed VS builds just project B. This non deterministic behavior where content of persistent cache depends on content of cross build process wide cache does not sounds right to me.

Reply:
The bad case here is just re-resolving a reference that we would have re-resolved with current behavior anyway—and afterwards, it can no longer be an issue for this assembly—so if we miss some cases, I'm not too worried about it. You're right that they can build on another process, but a lot of people use the same assemblies over and over again, so there's a decent chance most or all of the requisite assembly information is already in the cache anyway because of having built some other project that needed them.

Reply:
I measured total RAR time when building Ocelot with no live processes for each of noChange, currentChange, and addingMissingAssembliesToInstanceCache, once each without file caches and more with. Results:
Add missing files to instance cache: 10569.28 (no caches); 3276.099 (with cache)
Do not remove from instance cache: 9586.759 (no caches); 3264.925 (with cache)
Current version of MSBuild: 8569.355 (no caches); 6372.939

For the first run, writing all the files instead of just the ones a particular execution had had to figure out made the first version noticeably slower than the others. (I'm surprised there was so much of a difference between the version in this PR and main, but ah, well.) For incremental build, at least for what I ran, the difference seemed to be suspiciously huge, but I ran it several times, and it seemed pretty consistent. I'd appreciate a quick check on that anyway, since it does seem too good to be true. Anyway, the difference between adding everything to the instance cache versus just not removing anything seemed pretty small. I imagine that's a balance between slightly better cache performance and extra serializing/deserializing. Note that I was running a multiproc build for all tests.

From these results, keeping the PR as-is sounds optimal.

Reply:
Scenario which concerns me most is `hot rebuild`. That is either two rebuilds in max 15 minutes apart or more often rebuild max 15 minutes after incremental build. These scenarios are, IMHO, quite common as many developers, rightfully so, use rebuild as first remedy if something is fishy. Rebuild in such use cases will delete all caches, and since most of the data are in process-wide cache from previous builds it will create either very small cache file or none at all.

Also I believe that most common dev scenario is cold incremental - incremental build more apart than 15 minutes when there are no longer warm MSBuild  nodes.

I have measured it on orchard code. I have very fast SSD so on slower more common discs differences are expected to be bigger.
![image](https://user-images.githubusercontent.com/25249058/135605111-2b2a2367-9416-4cb2-98b8-0a4bcc7d63d6.png)
Scenario R - I - R - CI (rebuild, incremental, rebuild, cold incremental)  is where it is most noticeable. CI was measured by killing MSBuild processes before running incremental build.

Raw data:
Scenario | Type | RAR | Build | Cache size | Files
-- | -- | -- | -- | -- | --
R | no remove | 22644 | 34590 | 11755 | 338
R - I - I | no remove | 9640 | 6380 | 12363 | 345
R - I - CI | no remove | 12881 | 11020 | 14812 | 348
R - I - R - CI | no remove | 18461 | 12990 | 12636 | 349
R | old | 21872 | 37880 | 11614 | 337
R - I - I | old | 8648 | 6670 | 4989 | 341
R - I - CI | old | 20660 | 12420 | 11638 | 349
R - I - R - CI | old | 19820 | 13110 | 11876 | 348
R | add | 21684 | 34880 | 38105 | 353
R - I - I | add | 8451 | 6630 | 38105 | 353
R - I - CI | add | 11508 | 11840 | 38105 | 353
R - I - R - CI | add | 10996 | 12270 | 38105 | 353

As you can see on disc cache size (in kB) in `old` and `no-remove` is non deterministic but quite smaller (1 third). I value determinism in build processes - if not too costly.

Data shows that either solution is better than current one. Please feel free to pick one and proceed.

Reply:
My data suggested they were similar, but I didn't test all the scenarios you did, and add looks better than no remove for that, so I'll switch to add. Thanks for your help with this!

Reply:
Now I'm curious if I should check for the instance cache being up to date or just set it...but that sounds like a microoptimization, so I'm not going to worry about it.